rig/place_and_route/place/rcm.py

```diff
@@ -0,0 +1,182 @@
+    depth-first-search order.
+    """
+    if _visited is None:
+        _visited = set()
+
+    yield vertex
+    _visited.add(vertex)
+
+    for neighbour in vertices_neighbours[vertex]:
+        if neighbour not in _visited:
+            for vertex in _dfs(neighbour, vertices_neighbours, _visited):
+                yield vertex
+
+
+def _get_connected_subgraphs(vertices, vertices_neighbours):
+    """Break a graph containing unconnected subgraphs into a list of connected
+    subgraphs.
+
+    Returns
+    -------
+    [set([vertex, ...]), ...]
+    """
+    remaining_vertices = set(vertices)
+    subgraphs = []
+    while remaining_vertices:
+        subgraph = set(_dfs(remaining_vertices.pop(), vertices_neighbours))
+        remaining_vertices.difference_update(subgraph)
+        subgraphs.append(subgraph)
+
+    return subgraphs
+
+
+def _cuthill_mckee(vertices, vertices_neighbours):
+    """Yield the Cuthill-McKee order for a connected, undirected graph.
+
+    `Wikipedia
+    <https://en.wikipedia.org/wiki/Cuthill%E2%80%93McKee_algorithm>`_ provides
+    a good introduction to the Cuthill-McKee algorithm. The RCM algorithm
+    attempts to order vertices in a graph such that their adjacency matrix's
+    bandwidth is reduced. In brief the RCM algorithm is a breadth-first search
+    with the following tweaks:
+
+    * The search starts from the vertex with the lowest degree.
+    * Vertices discovered in each layer of the search are sorted by ascending
+      order of their degree in the output.
+
+    .. warning::
+
+        This function must not be called on a disconnected or empty graph.
+
+    Returns
+    -------
+    [vertex, ...]
+    """
+    vertices_degrees = {v: sum(itervalues(vertices_neighbours[v]))
+                        for v in vertices}
+
+    peripheral_vertex = min(vertices, key=(lambda v: vertices_degrees[v]))
+
+    visited = set([peripheral_vertex])
+    cm_order = [peripheral_vertex]
+    previous_layer = set([peripheral_vertex])
+    while len(cm_order) < len(vertices):
+        adjacent = set()
+        for vertex in previous_layer:
+            adjacent.update(vertices_neighbours[vertex])
+        adjacent.difference_update(visited)
+
+        visited.update(adjacent)
+        cm_order.extend(sorted(adjacent, key=(lambda v: vertices_degrees[v])))
+        previous_layer = adjacent
+
+    return cm_order
+
+
+def rcm_vertex_order(vertices_resources, nets):
+    """A generator which iterates over the vertices in Reverse-Cuthill-McKee
+    order.
+
+    For use as a vertex ordering for the sequential placer.
+    """
+    vertices_neighbours = _get_vertices_neighbours(nets)
+    for subgraph_vertices in _get_connected_subgraphs(vertices_resources,
+                                                      vertices_neighbours):
+        cm_order = _cuthill_mckee(subgraph_vertices, vertices_neighbours)
+        for vertex in reversed(cm_order):
+            yield vertex
```

Comment:
The `rcm_vertex_order` function's primary role is in fact for iterating over vertices in an application's graph. That said, as noted previously it is, in addition, internally used by `rcm_chip_order`.

Background: the `sequential` placer iterates over the vertices in an application graph and assigns them to chips in a machine, filling up one chip before moving on to the next. The order in which it iterates over vertices and fills chips is something the caller can control by providing an iterator over the vertices and chips respectively. The `breadth_first` placer simply provides a breadth-first-search-based iterator over vertices but doesn't specify a particular order in which to fill chips. The `hilbert` placer uses the same breadth-first vertex iterator used by the `breadth_first` placer but also provides an iterator over chips in a machine which causes the sequential placer to fill chips in hilbert-order.

The `rcm` placer simply defines that vertices should be placed in RCM order additionally that chips should be filled in RCM order. As such it implements two iterators `rcm_vertex_order` and `rcm_chip_order` which iterate over vertices in an application and chips in a machine (respectively) in RCM order. These iterators are then passed to the sequential placer which does the rest.